constraint-solver/src/solver/linearized.rs

```diff
@@ -0,0 +1,462 @@
+                // We always add the original constraint unmodified.
+                let mut constrs = vec![constr.clone()];
+                if !constr.is_affine() {
+                    let linearized = self.linearizer.linearize(
+                        constr,
+                        &mut || self.var_dispenser.next_var(),
+                        &mut constrs,
+                    );
+                    constrs.push(linearized);
+                }
+                constrs
+            })
+            .collect::<Vec<_>>();
+        self.solver.add_algebraic_constraints(constraints);
+    }
+
+    fn add_bus_interactions(
+        &mut self,
+        bus_interactions: impl IntoIterator<Item = BusInteraction<GroupedExpression<T, Variable<V>>>>,
+    ) {
+        let mut constraints_to_add = vec![];
+        let bus_interactions = bus_interactions
+            .into_iter()
+            .map(|bus_interaction| {
+                bus_interaction
+                    .fields()
+                    .map(|expr| {
+                        self.linearizer.substitute_by_var(
+                            expr.clone(),
+                            &mut || self.var_dispenser.next_var(),
+                            &mut constraints_to_add,
+                        )
+                    })
+                    .collect::<BusInteraction<_>>()
+            })
+            .collect_vec();
+        // We only substituted by a variable, but the substitution was not yet linearized.
+        self.add_algebraic_constraints(constraints_to_add);
+        self.solver.add_bus_interactions(bus_interactions);
+    }
+
+    fn add_range_constraint(
+        &mut self,
+        variable: &Variable<V>,
+        constraint: RangeConstraint<T::FieldType>,
+    ) {
+        self.solver.add_range_constraint(variable, constraint);
+    }
+
+    fn retain_variables(&mut self, variables_to_keep: &HashSet<Variable<V>>) {
+        // There are constraints that only contain `Variable::Linearized` that
+        // connect quadratic terms with the original constraints. We could try to find
+        // those, but let's just keep all of them for now.
+        let mut variables_to_keep = variables_to_keep.clone();
+        variables_to_keep.extend(self.var_dispenser.all_dispensed_vars());
+        self.solver.retain_variables(&variables_to_keep);
+    }
+
+    fn range_constraint_for_expression(
+        &self,
+        expr: &GroupedExpression<T, Variable<V>>,
+    ) -> RangeConstraint<T::FieldType> {
+        self.internalized_versions_of_expression(expr)
+            .fold(RangeConstraint::default(), |acc, expr| {
+                acc.conjunction(&self.solver.range_constraint_for_expression(&expr))
+            })
+    }
+
+    fn are_expressions_known_to_be_different(
+        &mut self,
+        a: &GroupedExpression<T, Variable<V>>,
+        b: &GroupedExpression<T, Variable<V>>,
+    ) -> bool {
+        let a = self.internalized_versions_of_expression(a);
```

Comment:
Now I'm actually not sure any more if this is really needed:
The base solver already searches for expressions equivalent to `a` by looking for algebraic constraints of the form `a' - a = 0`. And at least for direct substitutions, this should already be part of the constraint system. Although if we linearize, this could be a multi-step process and the solver might not be able to find it. Hm, maybe let's just keep it.